quantum/action_layer.c

```diff
@@ -287,6 +287,64 @@ uint8_t read_source_layers_cache(keypos_t key) {
 #    endif // ENCODER_MAP_ENABLE
     return 0;
 }
+
+#    ifdef KEYCODE_CACHE_ENABLE
+uint16_t keycode_map[((MATRIX_ROWS * MATRIX_COLS) + (CHAR_BIT)-1) / (CHAR_BIT)][16] = {{KC_NO}};
+#        ifdef ENCODER_MAP_ENABLE
+uint16_t encoder_keycode_map[(NUM_ENCODERS + (CHAR_BIT)-1) / (CHAR_BIT)][16] = {{KC_NO}};
+#        endif // ENCODER_MAP_ENABLE
```

Comment:
This looks overcomplicated.  The source layer cache code uses such tricks to avoid allocating a whole byte for each entry that needs just 2 bits in the most common case (4 layers); if you are storing the whole 16-bit keycode, you may just use
```c
static uint16_t keycode_map[MATRIX_ROWS][MATRIX_COLS];
```
and
```c
static uint16_t encoder_keycode_map[NUM_ENCODERS][NUM_DIRECTIONS];
```
(Actually the encoder code might not need such handling at all, because the press and release events are emitted without running the rest of main loop code in between, therefore keymap updates should not happen.)

Reply:
Thanks a ton for the feedback, I cleaned up the overcomplicated keycode_map and also removed the encoder handling